{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import subprocess\n",
    "import os\n",
    "from ucsc_genomes_downloader import download_genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "from math import floor, ceil\n",
    "def value_to_bed(base_pairs:int):\n",
    "    def _value_to_bed(value:str)->str:\n",
    "        chromosome, regions  = value.split(\"Mod_\")[1].split(\"_\")[0].split(\":\")\n",
    "        start, end = regions.split(\"-\")\n",
    "        start = int(start)\n",
    "        end = int(end)\n",
    "        delta = base_pairs - (end-start)\n",
    "        if delta < 0:\n",
    "            raise ValueError(\"Base pairs window is smaller than given file window.\")\n",
    "        start = start-int(floor(delta/2))\n",
    "        end = end+int(ceil(delta/2))\n",
    "        return f\"{chromosome},{start},{end},{chromosome}:{start}-{end}\"\n",
    "    return _value_to_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "def to_bed(df:pd.DataFrame, region:str, base_pairs:int)->pd.DataFrame:\n",
    "    df[region] = df[region].apply(value_to_bed(base_pairs))\n",
    "    df = pd.concat([\n",
    "        df[region].str.split(',', expand=True),\n",
    "        df.drop(columns=region)\n",
    "    ], axis=1)\n",
    "    df.columns = [\"chr\", \"start\", \"end\", \"id\", \"activity_ratio\"]\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['fastaFromBed', '-fi', 'hg19.fa', '-bed', 'liver_enhancers.bed', '-fo', 'liver_enhancers.fa'], returncode=0)"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_name = \"liver_enhancers\"\n",
    "genome = \"hg19\"\n",
    "base_pairs = 200\n",
    "region = f\"{file_name}.bed\"\n",
    "df = pd.read_csv(f\"{file_name}.tsv.gz\", sep=\"\\t\")\n",
    "download_genome(genome)\n",
    "to_bed(df, \"Region\", base_pairs).to_csv(region, index=False, sep=\"\\t\", header=False)\n",
    "goal = f\"{file_name}.fa\"\n",
    "subprocess.run([\n",
    "    \"fastaFromBed\",\n",
    "    \"-fi\",\n",
    "    f\"{genome}.fa\",\n",
    "    \"-bed\",\n",
    "    region,\n",
    "    \"-fo\",\n",
    "    goal\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
